### PR TITLE
Convert run_tests to python, allow running on Mac/Windows and allow filters for tests.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Sample usage:

To run only the embedder_unittests in the engine with the profile variant, the command would be
```
./flutter/testing/run_tests.py --variant host_profile_unopt --type engine --filter embedder_unittests
```

To run only the geometry in Dart with the debug variant, the command would be
```
./flutter/testing/run_tests.py --variant host_debug_unopt --type dart --filter geometry_test
```

Without any arguments, the behavior is identical to `run_tests.sh`.

In a subsequent patch, I will enable running unit-tests on Windows in the tryjobs. The lack of compatibility of the shell script on Windows made it so that we never ran any Windows unit-tests in the tryjobs.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
No change in functionality from the current version of the shell script. I will patch the tryjobs and remove the script next.

File: testing/run_tests.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  RunEngineExecutable(build_dir, 'ui_unittests', filter)
+
+  if IsMac():
+    RunEngineExecutable(build_dir, 'flutter_channels_unittests', filter)
+
+
+  if IsLinux():
+    RunEngineExecutable(build_dir, 'txt_benchmarks', filter, [ fonts_dir_flag ])
+
+  if IsLinux():
```
Comment (jason-simmons) [inline testing/run_tests.py]: Looks like this block was copied twice

Comment (chinmaygarde) [reply testing/run_tests.py]: Oh, one of them is a benchmark run.

File: testing/run_tests.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+  parser.add_argument('--variant', dest='variant', action='store', 
+      default='host_debug_unopt', help='The engine build variant to run the tests for.');
+  parser.add_argument('--type', type=str, choices=['all', 'engine', 'dart', 'benchmarks'], default='all')
+  parser.add_argument('--filter', type=str, default='',
+      help='The file name filter to use to select specific tests to run.')
+
+  args = parser.parse_args()
+  
+  run_engine_tests = args.type == 'engine'
```
Comment (liyuqian) [inline testing/run_tests.py]: Nit: `args.type in ['engine', 'all']` to save 4 lines below :)

File: testing/run_tests.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    if filter in os.path.basename(dart_test_file):
+      print "Testing dart file %s" % dart_test_file
+      RunDartTest(build_dir, dart_test_file, filter)
+    else:
+      print "Skipping %s due to filter." % dart_test_file
+
+
+def RunTests(build_dir, filter, run_engine_tests, run_dart_tests, run_benchmarks):
+  if run_engine_tests:
+    RunEngineTests(build_dir, filter)
```
Comment (liyuqian) [inline testing/run_tests.py]: Nit: the dart tests seem to be also engine tests. Shall we call this `RunGtests`, `RunCppTests`, `RunCcTests`, or just `RunUnitTests`?

**Comment by liyuqian** [comment]
BTW, I changed `—` to `--` in the PR description. You may need to manually change them during the merge. I guess the git commit somehow messed up with the copy/paste.

File: testing/run_tests.py (hunk 3 above)
Comment (chinmaygarde) [reply testing/run_tests.py]: Done. Used `RunCCTests`

File: testing/run_tests.py (hunk 2 above)
Comment (chinmaygarde) [reply testing/run_tests.py]: Cool. Done.

**Pull request merged**